lib/parse/parser_interpreter.es6.js

```diff
@@ -0,0 +1,299 @@
+
+      const browserName = match[1] && match[1].toLowerCase();
+      const browserVersion = match[2] && match[2].toLowerCase();
+      const osName = match[3] && match[3].toLowerCase();
+      const osVersion = match[4] && match[4].toLowerCase();
+
+      const columns = this.selectable
+            .filter(col => this.CompatProperty.isInstance(col));
+
+      return columns.filter(
+          c => (!browserName || c.release.browserName.toLowerCase()
+                .startsWith(browserName)) &&
+            (!browserVersion || c.release.browserVersion.toLowerCase()
+             .startsWith(browserVersion)) &&
+            (!osName || c.release.osName.toLowerCase()
+             .startsWith(osName)) &&
+            (!osVersion || c.release.osVersion.toLowerCase()
+             .startsWith(osVersion)));
+    },
+  ],
+});
+
+foam.CLASS({
+  name: 'QueryParser',
+  package: 'org.chromium.parse',
+  implements: ['foam.mlang.Expressions'],
+
+  requires: [
+    'foam.core.Property',
+    'foam.core.String',
+    'org.chromium.parse.DefaultQueryInterpreter',
+    'org.chromium.parse.ReleaseApiQueryInterpreter',
+  ],
+
+  properties: [
+    {
+      class: 'Class',
+      name: 'of',
+      documentation: 'The class of items in the DAO being queried.',
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'org.chromium.parse.QueryInterpreter',
+      name: 'interpreter',
+      factory: function() {
+        return this.ReleaseApiQueryInterpreter.create({
+          delegate: this.DefaultQueryInterpreter.create(),
+        });
+      },
+    },
+  ],
+
+  methods: [
+    function parseString(str, opt_name) {
+      return this.queryGrammar.parseString(str, opt_name);
+    },
+  ],
+
+  grammars: [
+    {
+      name: 'queryGrammar',
+      language: 'foam.parse.Parsers',
+      symbols: function() {
+        return {
+          START: sym('query'),
+          query: seq1(1, sym('ws'), sym('or'), sym('ws')),
+
+          or: plus(sym('and'), sym('orDelim')),
+          orDelim: alt(seq(sym('ws'), literalIC('OR'), sym('ws')),
+                       seq(sym('ws'), '|', sym('ws'))),
+
+          and: plus(sym('expr'), sym('andDelim')),
+          andDelim: alt(seq(sym('ws'), literalIC('AND'), sym('ws')),
+                        not(sym('orDelim'), sym('wsRequired'))),
+
+          expr: alt(sym('paren'),
+                    sym('keyValue'),
+                    sym('keyword')),
+
+          paren: seq1(1, '(', sym('query'), ')'),
+
+          keyValue: seq(sym('key'), ':', sym('value')),
+          key: str(plus(alt(range('a', 'z'), range('A', 'Z'),
+                            range('0', '9'), '.', '_'))),
+          value: str(plus(alt(range('a', 'z'), range('A', 'Z'),
+                              range('0', '9'), '.', '_', '#', ','))),
+
+          keyword: str(plus(alt(range('a', 'z'), range('A', 'Z'),
+                                range('0', '9'), '.', '_', '#'))),
+
+          ws: repeat0(chars(' \t\r\n')),
```

Comment:
Would the grammar be any less complicated if you collapsed whitespace before feeding it to the parser? Since you don't have to deal with strings in your query language, collapsing them to a single space should be easy.

Reply:
I'm not convinced it would be substantially simpler. We would essentially replace `sym('ws')` with `optional(' ')`.